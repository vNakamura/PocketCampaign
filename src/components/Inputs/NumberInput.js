// @flow

import React, {Component} from 'react';
import styled from 'styled-components';

import Input from './Input';
import Button from '../Buttons/Button';

const Container = styled.div `
  display: flex;
  flex-direction: column;
  align-items: stretch;
`;

type Props = {
  defaultValue: number,
  step: number,
  min: number,
  max: number,
  onChange?: (value: number) => void,
  hideSign: bool,
};

export default class NumberInput extends Component {
  props: Props;

  state: {
    value: number,
  };

  constructor(props: Props) {
    super(props);
    this.state = {
      value: props.defaultValue,
    };
  };

  static defaultProps = {
    defaultValue: 0,
    step: 1,
    min: Number.MIN_SAFE_INTEGER || -9007199254740991,
    max: Number.MAX_SAFE_INTEGER || 9007199254740991,
    hideSign: false,
  };

  static filterInt(value: string | number): number {
    const adjustedValue = String(value);
    if (/^(-|\+)?([0-9]+)$/.test(adjustedValue)) return Number(adjustedValue);
    return 0;
  };

  applyValue = (value: string | number, modifier: number = 0): void => {
    const { max, min, onChange } = this.props;
    let adjustedValue = NumberInput.filterInt(value) + modifier;
    adjustedValue = Math.min(max, adjustedValue);
    adjustedValue = Math.max(min, adjustedValue);
    this.setState({ value: adjustedValue });
    if (onChange) onChange(adjustedValue);
  };

  handleChange = (e: Event): void => {
    const target: EventTarget = e.currentTarget;
    if(target instanceof HTMLInputElement) this.applyValue(target.value);
  };

  handleKeyDown = (e: KeyboardEvent): void => {
    const { step } = this.props;
    if (e.keyCode === 38) { // Up
      e.preventDefault();
      this.applyValue(
        this.state.value,
        e.ctrlKey || e.metaKey || e.shiftKey ? 10 * step : step,
      );
    } else if (e.keyCode === 40) { // Down
      e.preventDefault();
      this.applyValue(
        this.state.value,
        e.ctrlKey || e.metaKey || e.shiftKey ? -10 * step : -step,
      );
    }
  };

  increase = (): void => {
    this.applyValue(this.state.value, this.props.step);
  };

  decrease = (): void => {
    this.applyValue(this.state.value, -this.props.step);
  };

  render() {
    const { hideSign } = this.props;
    const value = hideSign ? Math.abs(this.state.value) : this.state.value;
    return (
      <Container>
        <Button
          onClick={this.increase}
          tabIndex="-1"
        >+</Button>
        <Input
          value={value}
          onChange={this.handleChange}
          onKeyDown={this.handleKeyDown}
          size={value.toString().length}
        />
        <Button
          onClick={this.decrease}
          tabIndex="-1"
        >-</Button>
      </Container>
    );
  }
};
