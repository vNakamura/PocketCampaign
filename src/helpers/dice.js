// @flow

import React from 'react';
import droll from 'droll';
import seedrandom from 'seedrandom';
import replace from 'string-replace-to-array';

import NotationLink from '../components/Chat/NotationLink';

export const notationLimits = {
  minNumDice: 1,
  maxNumDice: 10,
  minNumSides: 2,
  maxNumSides: 100,
  minModifier: -1000,
  maxModifier: 1000,
};
export const diceNotationRegex = /\b([1-9]\d*)?d([1-9]\d*)([+-]\d+)?\b/gi;

type ResultError = {
  ok: false,
  message: string,
};
type ValidateResult = ResultError | { ok: true };
const validateRoll = (notation:string): ValidateResult => {
  const p = droll.parse(notation);
  if (!p) return { ok: false, message: 'Invalid dice notation' };
  if (p.numDice > notationLimits.maxNumDice) return { ok: false, message: 'Invalid dice notation. Too many dice!' };
  if (p.numSides < notationLimits.minNumDice) return { ok: false, message: 'Invalid dice notation. You need at least one die!' };
  if (p.numSides > notationLimits.maxNumSides) return { ok: false, message: 'Invalid dice notation. Too many sides!' };
  if (p.numSides < notationLimits.minNumSides) return { ok: false, message: 'Invalid dice notation. Too few sides!' };
  if (p.modifier > notationLimits.maxModifier) return { ok: false, message: 'Invalid dice notation. Modifier is too big!' };
  if (p.modifier < notationLimits.minModifier) return { ok: false, message: 'Invalid dice notation. Modifier is too big!' };
  return { ok: true };
};

export type ParseResult = ResultError | {
  ok: true,
  parsed: {
    numDice: number,
    numSides: number,
    modifier: number,
    minResult: number,
    maxResult: number,
    avgResult: number,
  }
};
export const parseRoll = (notation: string): ParseResult => {
  const v = validateRoll(notation);
  if (!v.ok) return v;
  const p = droll.parse(notation);
  return { ok: true, parsed: p };
};

export type RollResult = ResultError | {
  ok: true,
  rolls: number[],
  modifier: number,
  total: number,
};
export const roll = (notation: string, seed?: string): RollResult => {
  const v = validateRoll(notation);
  if (!v.ok) return v;

  if (seed !== undefined) seedrandom(seed, { global: true });
  const result = droll.roll(notation);
  return {
    ok: true,
    rolls: result.rolls,
    modifier: result.modifier,
    total: result.total,
  };
};

export const addLinksToText = (text: string) => {
  let i = 0;
  const result = replace(
    text,
    diceNotationRegex,
    (notation, numDice, numSides) => {
      i += 1;
      const v: ValidateResult = validateRoll(notation);
      if (v.ok && numSides > 1) return <NotationLink key={i} notation={notation} />;
      return notation;
    },
  );
  return result;
};
