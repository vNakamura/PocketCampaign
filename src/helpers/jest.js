import React from 'react';
import { ThemeProvider } from 'styled-components';
/* eslint-disable import/no-extraneous-dependencies */
import renderer from 'react-test-renderer';
import 'jest-styled-components';
/* eslint-enable import/no-extraneous-dependencies */
import theme from '../theme';

export const applyTheme = content => (
  <ThemeProvider theme={theme}>
    {content}
  </ThemeProvider>
);

export const renderWithTheme = content => renderer.create(
  applyTheme(content),
).toJSON();

export default {
  applyTheme,
  renderWithTheme,
};
