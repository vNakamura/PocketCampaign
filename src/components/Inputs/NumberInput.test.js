import React from 'react';
import { mount, shallow } from 'enzyme';

import NumberInput from './NumberInput';
import { applyTheme, renderWithTheme } from '../../helpers/jest';

const changeHandler = jest.fn();

describe('components: Inputs/NumberInput', () => {
  it('should render correctly', () => {
    const signed = renderWithTheme(<NumberInput hideSign />);
    expect(signed).toMatchSnapshot();
    const NotSigned = renderWithTheme(<NumberInput />);
    expect(NotSigned).toMatchSnapshot();
  });

  it('should change value when buttons are pressed', () => {
    const wrapper = mount(applyTheme(<NumberInput
      defaultValue={5}
      step={2}
      onChange={changeHandler}
    />));
    const buttons = wrapper.find('button');
    changeHandler.mockClear();
    buttons.at(0).simulate('click');
    expect(changeHandler).toHaveBeenCalledWith(7);
    changeHandler.mockClear();
    buttons.at(1).simulate('click');
    expect(changeHandler).toHaveBeenCalledWith(5);
  });

  it('should change value when up and down keys are pressed', () => {
    const wrapper = mount(applyTheme(<NumberInput
      defaultValue={8}
      step={3}
      onChange={changeHandler}
    />));
    const input = wrapper.find('input');
    changeHandler.mockClear();
    input.simulate('keyDown', { keyCode: 38 });
    expect(changeHandler).toHaveBeenCalledWith(11);
    changeHandler.mockClear();
    input.simulate('keyDown', { keyCode: 40 });
    expect(changeHandler).toHaveBeenCalledWith(8);
  });

  it('should change value if defaultValue prop changes', () => {
    const wrapper = shallow(applyTheme(<NumberInput
      defaultValue={30}
    />)).find(NumberInput).dive();
    expect(wrapper.state('value')).toBe(30);
    wrapper.setProps({ defaultValue: 97 });
    expect(wrapper.state('value')).toBe(97);
  });

  it('should accept only numeric values', () => {
    expect(NumberInput.filterInt(3)).toBe(3);
    expect(NumberInput.filterInt('100')).toBe(100);
    expect(NumberInput.filterInt('-20')).toBe(-20);
    expect(NumberInput.filterInt('a')).toBe(0);
    expect(NumberInput.filterInt('a1')).toBe(0);
  });
});
