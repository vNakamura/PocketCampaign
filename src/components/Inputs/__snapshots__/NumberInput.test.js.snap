// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components: Inputs/NumberInput should render correctly 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  padding: .2em;
}

.c1 {
  font-size: 1em;
  margin: 0;
  padding: .1em .4em;
}

.c2 {
  font-family: "Open Sans";
  font-size: 1rem;
  padding: .4em;
  border: none;
  border-radius: 0;
  text-align: center;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
}

.c4 {
  background-color: #e6e5f1;
  color: #2a2932;
  margin: 0;
  padding: .1em .4em;
}

.c4:hover {
  background-color: #e05858;
}

.c5 {
  background: none;
  border: none;
  cursor: pointer;
  color: #817b9e;
  font-family: Dosis;
  font-size: 1rem;
  padding: .2em .4em;
}

.c5:focus {
  outline: none;
}

.c5:hover,
.c5:focus {
  color: #e6e5f1;
}

<div
  className="c0"
>
  <input
    className="c1 c2"
    onChange={[Function]}
    onKeyDown={[Function]}
    size={1}
    type="tel"
    value={0}
  />
  <div
    className="c3"
  >
    <button
      className="c4 c5"
      onClick={[Function]}
      onMouseUp={[Function]}
      tabIndex="-1"
    >
      <svg
        fill="currentColor"
        height="1em"
        preserveAspectRatio="xMidYMid meet"
        style={
          Object {
            "color": undefined,
            "verticalAlign": "middle",
          }
        }
        viewBox="0 0 40 40"
        width="1em"
      >
        <g>
          <path
            d="m31 26.4q0 0.3-0.2 0.5l-1.1 1.2q-0.3 0.2-0.6 0.2t-0.5-0.2l-8.7-8.8-8.8 8.8q-0.2 0.2-0.5 0.2t-0.5-0.2l-1.2-1.2q-0.2-0.2-0.2-0.5t0.2-0.5l10.4-10.4q0.3-0.2 0.6-0.2t0.5 0.2l10.4 10.4q0.2 0.2 0.2 0.5z"
          />
        </g>
      </svg>
    </button>
    <button
      className="c4 c5"
      onClick={[Function]}
      onMouseUp={[Function]}
      tabIndex="-1"
    >
      <svg
        fill="currentColor"
        height="1em"
        preserveAspectRatio="xMidYMid meet"
        style={
          Object {
            "color": undefined,
            "verticalAlign": "middle",
          }
        }
        viewBox="0 0 40 40"
        width="1em"
      >
        <g>
          <path
            d="m31 16.4q0 0.3-0.2 0.5l-10.4 10.4q-0.3 0.3-0.5 0.3t-0.6-0.3l-10.4-10.4q-0.2-0.2-0.2-0.5t0.2-0.5l1.2-1.1q0.2-0.2 0.5-0.2t0.5 0.2l8.8 8.8 8.7-8.8q0.3-0.2 0.5-0.2t0.6 0.2l1.1 1.1q0.2 0.2 0.2 0.5z"
          />
        </g>
      </svg>
    </button>
  </div>
</div>
`;

exports[`components: Inputs/NumberInput should render correctly 2`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  padding: .2em;
}

.c1 {
  font-size: 1em;
  margin: 0;
  padding: .1em .4em;
}

.c2 {
  font-family: "Open Sans";
  font-size: 1rem;
  padding: .4em;
  border: none;
  border-radius: 0;
  text-align: center;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
}

.c4 {
  background-color: #e6e5f1;
  color: #2a2932;
  margin: 0;
  padding: .1em .4em;
}

.c4:hover {
  background-color: #e05858;
}

.c5 {
  background: none;
  border: none;
  cursor: pointer;
  color: #817b9e;
  font-family: Dosis;
  font-size: 1rem;
  padding: .2em .4em;
}

.c5:focus {
  outline: none;
}

.c5:hover,
.c5:focus {
  color: #e6e5f1;
}

<div
  className="c0"
>
  <input
    className="c1 c2"
    onChange={[Function]}
    onKeyDown={[Function]}
    size={1}
    type="tel"
    value={0}
  />
  <div
    className="c3"
  >
    <button
      className="c4 c5"
      onClick={[Function]}
      onMouseUp={[Function]}
      tabIndex="-1"
    >
      <svg
        fill="currentColor"
        height="1em"
        preserveAspectRatio="xMidYMid meet"
        style={
          Object {
            "color": undefined,
            "verticalAlign": "middle",
          }
        }
        viewBox="0 0 40 40"
        width="1em"
      >
        <g>
          <path
            d="m31 26.4q0 0.3-0.2 0.5l-1.1 1.2q-0.3 0.2-0.6 0.2t-0.5-0.2l-8.7-8.8-8.8 8.8q-0.2 0.2-0.5 0.2t-0.5-0.2l-1.2-1.2q-0.2-0.2-0.2-0.5t0.2-0.5l10.4-10.4q0.3-0.2 0.6-0.2t0.5 0.2l10.4 10.4q0.2 0.2 0.2 0.5z"
          />
        </g>
      </svg>
    </button>
    <button
      className="c4 c5"
      onClick={[Function]}
      onMouseUp={[Function]}
      tabIndex="-1"
    >
      <svg
        fill="currentColor"
        height="1em"
        preserveAspectRatio="xMidYMid meet"
        style={
          Object {
            "color": undefined,
            "verticalAlign": "middle",
          }
        }
        viewBox="0 0 40 40"
        width="1em"
      >
        <g>
          <path
            d="m31 16.4q0 0.3-0.2 0.5l-10.4 10.4q-0.3 0.3-0.5 0.3t-0.6-0.3l-10.4-10.4q-0.2-0.2-0.2-0.5t0.2-0.5l1.2-1.1q0.2-0.2 0.5-0.2t0.5 0.2l8.8 8.8 8.7-8.8q0.3-0.2 0.5-0.2t0.6 0.2l1.1 1.1q0.2 0.2 0.2 0.5z"
          />
        </g>
      </svg>
    </button>
  </div>
</div>
`;
